src/postgraphile/http/subscriptions.ts

```diff
@@ -186,134 +184,278 @@ export async function enhanceHttpServerWithSubscriptions<
-
-          return response;
-        };
-        // onOperation is only called once per params object, so there's no race condition here
-        // eslint-disable-next-line require-atomic-updates
-        params.formatResponse = formatResponse;
-        const hookedParams = pluginHook
-          ? pluginHook('postgraphile:ws:onOperation', params, {
-              message,
-              params,
-              socket,
+            // for supplying custom execution arguments
+            const hookedArgs = pluginHook
+              ? pluginHook('postgraphile:ws:onSubscribe', args, {
+                  ...ctx,
+                  message: msg,
+                  options,
+                })
+              : args;
+
+            // You are strongly encouraged to use
+            // `postgraphile:validationRules:static` if possible - you should
+            // only use this one if you need access to variables.
+            const { req, res } = await reqResFromSocket(ctx.socket);
+            const validationRules = pluginHook('postgraphile:validationRules', [], {
               options,
-            })
-          : params;
-        const finalParams: typeof hookedParams & { query: DocumentNode } = {
-          ...hookedParams,
-          query:
-            typeof hookedParams.query !== 'string' ? hookedParams.query : parse(hookedParams.query),
-        };
-
-        // You are strongly encouraged to use
-        // `postgraphile:validationRules:static` if possible - you should
-        // only use this one if you need access to variables.
-        const moreValidationRules = pluginHook('postgraphile:validationRules', [], {
-          options,
-          req,
-          res,
-          variables: params.variables,
-          operationName: params.operationName,
-          meta,
-        });
-        if (moreValidationRules.length) {
-          const validationErrors: ReadonlyArray<GraphQLError> = validate(
-            params.schema,
-            finalParams.query,
-            moreValidationRules,
-          );
-          if (validationErrors.length) {
-            const error = new Error(
-              'Query validation failed: \n' + validationErrors.map(e => e.message).join('\n'),
+              req,
+              res,
+              variables: hookedArgs.variableValues,
+              operationName: hookedArgs.operationName,
+              // no meta because validation errors returned from here will be
+              // served through the error message. it contains just the GraphQLErrors
+              // (there is no result to add the meta to)
+            });
+            const validationErrors = validate(
+              hookedArgs.schema,
+              hookedArgs.document,
+              // should the validation rules be applied
+              // in addition to the graphql defaults (as before)
+              // or should they be used exclusively?
+              validationRules.length ? validationRules : undefined,
             );
```

Comment:
The query should already have been validated at this point, right?

If so, we only want to apply _additional_ validation rules here. So there should be an `if (moreValidationRules.length) {` check around this.

If not, we need to be much more careful with the validation here - at the moment adding a single validation rule will cause all the built in validations not to be used.

Reply:
When supplying custom `onSubscribe` execution arguments, [you're trusted to prepare the execution arguments](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/_server_.serveroptions.md#onsubscribe) for direct usage. This is a conscious decision because it allows you to apply optimizations (like executing some rules exclusively or trusting the client implicitly) or simply have control of when the validation happens when you introduce a custom implementation.

> The query should already have been validated at this point, right?

So no, at this point the query is untouched. As you may notice, we're using the fields from the raw message.

> If not, we need to be much more careful with the validation here - at the moment adding a single validation rule will cause all the built in validations not to be used.

Yeah, which is why I asked. Didn't really know how strict the hook is. [I now do the regular validation before performing optional additional ones.](https://github.com/graphile/postgraphile/pull/1338/commits/4639fcc1259ae05ab9b7f91bc1f4a86c4eb2d810)

Reply:
Got it. So there's a bit more work you need to do; your validation needs to use the static rules, like here:

https://github.com/graphile/postgraphile/blob/7dde41da9204bbd381c20e5783439118650fccd2/src/postgraphile/http/createPostGraphileHttpRequestHandler.ts#L381

which come from:

https://github.com/graphile/postgraphile/blob/7dde41da9204bbd381c20e5783439118650fccd2/src/postgraphile/http/createPostGraphileHttpRequestHandler.ts#L324-L326

(Note that they're based on GraphQL's default list of validation rules, but can then be manipulated via server plugins.)

(Further, note these rules only need to be calculated once, e.g. within the `enhanceHttpServerWithWebSockets` call. They do not need to be recalculated for each request.)

Then later comes the part I think you've now handled; i.e.

https://github.com/graphile/postgraphile/blob/7dde41da9204bbd381c20e5783439118650fccd2/src/postgraphile/http/createPostGraphileHttpRequestHandler.ts#L820-L838

Reply:
Awesome! [The `staticValidationRules` were already available for me, I just plugged them in.](https://github.com/graphile/postgraphile/pull/1338/commits/f78a4cc4d4ca55bd8d2701fab74d1d8d8826bf8e)